### PR TITLE
test(redteam): complete_empty_exam_session attack-surface spec (#557)

### LINKED ISSUE #557: Red-team specs for complete_empty_exam_session attack surface
## Why
Commit `375bde5` adds RPC `complete_empty_exam_session` and Server Action `submitEmptyExamSession`. Red-team agent flagged 4 new attack vectors with no spec coverage:

- **AN (HIGH)** — `submitEmptyExamSession` called without auth.
- **AO (HIGH)** — IDOR: student calls RPC with another student's session_id.
- **AP (HIGH)** — Mode confusion: student passes a study session_id; RPC must reject.
- **AQ (MEDIUM)** — Cross-RPC idempotency: `batch_submit_quiz` writes a real score, then `complete_empty_exam_session` is invoked; verify final state is the real score, not 0%.

## What
1. New `apps/web/e2e/redteam/rpc-complete-empty-exam.spec.ts` — covers AO, AP, AQ.
2. Add a test case to `apps/web/e2e/redteam/server-action-unauthenticated.spec.ts` — covers AN.

## Acceptance
- `pnpm --filter @repo/web e2e:redteam` passes with the new spec(s).
- All 4 vectors execute and assert the expected RPC/action error codes.

Linked: PR #523 round 7 Phase 1.

Priority: P1. Size: S.

### PR DESCRIPTION
**Pull request by okpilot**
## What

New red-team Playwright spec for the `complete_empty_exam_session` SECURITY DEFINER RPC (sibling of `complete_overdue_exam_session`). **Coverage-only — no production code change.**

## Vectors covered (attack-surface.md GAP → COVERED)

| Vector | Attack | Expected defense |
|--------|--------|------------------|
| **AO** (HIGH) | different student completes a foreign session (IDOR) | `session not found or not accessible` |
| **AP** (HIGH) | owner completes a non-exam (`quick_quiz`) session | `session is not an exam` |
| **AQ** (MED) | idempotent replay of an already-scored session | re-reads real DB score (75/true **and** sub-pass 50/false variants); never clobbers to 0/false |
| **AR** (LOW) | owner completes a soft-deleted session | `session not found or not accessible` |

Plus: **AO+AQ** (a non-owner can't replay a foreign *completed* session — ownership fires before the idempotent branch), soft-deleted-user guard (`user not found or inactive`), and positive paths for `mock_exam`, `internal_exam`, and a **not-yet-overdue early-submit** session (the key behavioral difference from `complete_overdue`).

## Verification

- Assertions verified against the latest RPC definition (mig `20260429000008`); guard order + AQ idempotent re-read confirmed by impl-critic + semantic-reviewer.
- biome / tsc / build green; CR local triaged (full output-contract assertions added; one finding skipped — `quiz_sessions` has no `answered_count` column, the RPC counts `quiz_session_answers`).
- Admin assertion reads destructure `{ error }` per code-style.md §5. Redteam Playwright suite runs in CI.

## Notes

- AN (unauthenticated) deferred to the Hub A `server-action-unauthenticated.spec.ts`.

Closes #557

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Added a comprehensive red-team end-to-end spec for completing exam sessions, validating security, idempotency, replay behavior, boundary cases and cleanup.

* **Documentation**
  * Updated testing coverage and attack-surface tracking to reflect the new spec and corrected descriptions.
  * Added recurring notes and a learner rule candidate emphasizing full-output assertions and idempotent/replay test patterns.

### CODE REVIEW

File: apps/web/e2e/redteam/rpc-complete-empty-exam.spec.ts [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+      // Log rather than throw — a throw here would mask a try-block assertion
+      // failure (noUnsafeFinally).
+      const { error: restoreErr } = await admin
+        .from('users')
+        .update({ deleted_at: null })
+        .eq('id', victimUserId)
+      if (restoreErr) {
+        console.error(`[complete-empty] restore victim user failed: ${restoreErr.message}`)
+      }
+    }
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/e2e/redteam/rpc-complete-empty-exam.spec.ts:349]: Fixed in 5ad6be8f — the restore result is now captured in an outer-scoped `restoreErrorMessage` and `expect(...).toBeNull()` is asserted after the try/finally, so a failed restore fails the test (without throwing inside finally). Thanks — good catch on the cross-spec contamination risk.
